targets/mbedos5/tools/generate_pins.py

```diff
@@ -0,0 +1,224 @@
+
+from simpleeval import SimpleEval, DEFAULT_OPERATORS
+
+import ast
+
+import argparse
+import json
+import sys
+import os
+
+# import mbed tools
+sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'mbed-os'))
+from tools.targets import Target
+
+
+def find_file(root_dir, directories, name):
+    """
+    Find the first instance of file with name 'name' in the directory tree
+    starting with 'root_dir'.
+
+    Filter out directories that are not in directories, or do not start with
+    TARGET_.
+
+    Since this looks in (essentially )the same directories as the compiler would
+    when compiling mbed OS, we should only find one PinNames.h.
+    """
+
+    for root, dirs, files in os.walk(root_dir, topdown=True):
+        # modify dirs in place
+        dirs[:] = filter(lambda x: x in directory_labels or not x.startswith('TARGET_'), dirs)
+
+        if name in files:
+            return os.path.join(root, name)
+
+def enumerate_includes(root_dir, directories):
+    """
+    Walk through the directory tree, starting at root_dir, and enumerate all
+    valid include directories.
+    """
+    for root, dirs, files in os.walk(root_dir, topdown=True):
+        # modify dirs in place
+        dirs[:] = filter(lambda x: x in directory_labels 
+                                or (    not x.startswith('TARGET_')
+                                    and not x.startswith('TOOLCHAIN_')), dirs)
+        yield root
+
+
+class TypeDeclVisitor(c_ast.NodeVisitor):
+    def __init__(self, filter_names=[]):
+        self.names = filter_names
+
+    def visit(self, node):
+        value = None
+
+        if node.__class__.__name__ == "TypeDecl":
+            value = self.visit_TypeDecl(node)
+
+        if value is None:
+            for name, c in node.children():
+                value = value or self.visit(c)
+
+        return value
+
+    def visit_TypeDecl(self, node):
+        if node.declname in self.names:
+            c_gen = c_generator.CGenerator()
+            pins = {}
+
+            operators = DEFAULT_OPERATORS
+            operators[ast.BitOr] = lambda a, b: a | b
+            operators[ast.LShift] = lambda a, b: a << b
+            operators[ast.RShift] = lambda a, b: a << b
+            evaluator = SimpleEval(DEFAULT_OPERATORS )
+
+            for pin in node.type.values.enumerators:
+                expr = c_gen.visit(pin.value)
+
+                if "(int)" in expr:
+                    expr = expr.replace('(int)', '')
+
+                if expr in pins:
+                    pins[pin.name] = pins[expr]
+                else:
+                    pins[pin.name] = evaluator.eval(expr.strip())
+
+            return pins
+
+def enumerate_pins(c_source_file, include_dirs, definitions):
+    """
+    Enumerate pins specified in PinNames.h, by looking for a PinName enum
+    typedef somewhere in the file.
+    """
+    definitions += ['__attribute(x)__=', '__extension__(x)=', 'register=', '__IO=', 'uint32_t=unsigned int']
```

Comment:
Could we merge this list with the gcc_args below? That way we don't need to modify the `definitions` list and also no need for the extra list copy before calling this method.